File: Domain/ListingAggregate/Listing.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Domain.CompanyAggregate;
using Domain.ListingAggregate.Enums;
using Domain.ListingAggregate.Objects;
using Domain.Enums;
using Domain.MediaAggregate;

namespace Domain.ListingAggregate
{
    public class Listing
    {
        public Guid Id { get; set; }
        public bool Accessibility { get; set; }
        public DateTime AddedOn { get; set; }
        public string AdministrationFees { get; set; }
        public double AnnualBusinessRates { get; set; }
        public double AreaTotal { get; set; }
        public UnitOfArea AreaUnits { get; set; }
        public AccessStatus AccessStatus { get; set; }
        public int AvailableBedrooms { get; set; }
        public DateTime AvailableFromDate { get; set; }
        public int Bathrooms { get; set; }
        public List<Utility> BillsIncluded { get; set; }
        public bool BusinessForSale { get; set; }
        public List<Incentive> BuyerIncentives { get; set; }
        public Category Category { get; set; }
        public CentralHeating CentralHeating { get; set; }
        public bool ChainFree { get; set; }
        public List<string> CommercialUseClass { get; set; }
        public string CommonholdDetails { get; set; }
        public List<Utility> ConnectedUtilities { get; set; }
        public int ConstructionYear { get; set; }
        public ICollection<ListingMedia> ListingMedia { get; set; }
        public CookerType CookerType { get; set; }
        public CouncilTaxBand CouncilTaxBand { get; set; }
        public DecorativeCondition DecorativeCondition { get; set; }
        public double Deposit { get; set; }
        public ICollection<DetailedDescription> DetailedDescriptions { get; set; }
        public string EerCurrentRating { get; set; }
        public string EerPotentialRating { get; set; }
        public string EirCurrentRating { get; set; }
        public string EirPotentialRating { get; set; }
        public bool FeatureProperty { get; set; }
        public List<string> FeatureList { get; set; }
        public List<string> FloorLevels { get; set; }
        public int Floors { get; set; }
        public FurnishedState FurnishedState { get; set; }
        public double GroundRent { get; set; }
        public string GroundRentReviewPeriod { get; set; }
        public string LeaseExpiry { get; set; }
        public LifeCycleStatus LifeCycleStatus { get; set; }
        public bool ListedBuilding { get; set; }
        public ListedBuildingGrade ListedBuildingGrade { get; set; }
        public string ListingReference { get; set; }
        public ListingLocation ListingLocation { get; set; }
        public int LivingRooms { get; set; }
        public int MinimumContractLength { get; set; }
        public UnitOfTime MinimumContractLengthUnits { get; set; }
        public bool NewBuild { get; set; }
        public DateTime OpenDay { get; set; }
        public List<FeatureSpace> FeatureSpaces { get; set; }
        public List<Parking> Parking { get; set; }
        public bool PetsAllowed { get; set; }
        public Pricing Pricing { get; set; }
        public PropertyType PropertyType { get; set; }
        public int RateableValue { get; set; }
        public int RatesPayable { get; set; }
        public RentalTerm RentalTerm { get; set; }
        public bool Repossession { get; set; }
        public bool Retirement { get; set; }
        public int SapRating { get; set; }
        public ServiceCharge ServiceCharge { get; set; }
        public bool Serviced { get; set; }
        public bool SharedAccommodation { get; set; }
        public string SharedOwnershipDetails { get; set; }
        public bool SmokersConsidered { get; set; }
        public string SummaryDescription { get; set; }
        public bool Tenanted { get; set; }
        public Eligibility TenantEligibilityDss { get; set; }
        public Eligibility TenantEligibilityStudents { get; set; }
        public Tenure Tenure { get; set; }
        public int TotalBedrooms { get; set; }
        public List<UniqueFeature> UniqueFeatures { get; set; }
        public List<WhiteGoods> WhiteGoods { get; set; }
        public Guid CompanyId { get; set; }
        public Company Company { get; set; }
        public ICollection<ListingWatcher> Watchers { get; set; } = new List<ListingWatcher>();
    }
}

File: client-app/src/app/model/ListingAggregate/Listing.ts
import { AccessStatus } from "../AccessStatus";
import { UnitOfTime } from "../Membership";
import {
    Category, CentralHeating, CookerType, CouncilTaxBand, DecorativeCondition, 
    Eligibility, FurnishedState, 
    LifeCycleStatus, ListedBuildingGrade, PropertyType, 
    RentalTerm, Tenure, UnitOfArea, TransactionType, Frequency, 
} from "./ListingEnums";
import { ChangeLog, DetailedDescription, KeyContact, ListingLocation, 
    ListingMediaDto, 
    Owner, Pricing, ServiceCharge, WatcherDto, } from "./ListingObjects";

export interface Listing {
    id: string;
    accessibility: boolean;
    accessStatus: AccessStatus;
    acreage: number;
    agency: string;
    addedOn: Date;
    administrationFees: string;
    annualBusinessRates: number;
    areaTotal: number;
    areaUnits: UnitOfArea;
    availableBedrooms: number;
    availableFromDate: Date;
    bathrooms: number;
    billsIncluded: string[];
    businessForSale: boolean;
    buyerIncentives: string[];
    company: Owner;
    companyId: string;
    category: Category;
    centralHeating: CentralHeating;
    chainFree: boolean;
    changeLogs: ChangeLog[];
    commercialUseClass: string[];
    commonholdDetails: string;
    connectedUtilities: string[];
    constructionYear: number;
    cookerType: CookerType;
    councilTaxBand: CouncilTaxBand;
    decorativeCondition: DecorativeCondition;
    deposit: number;
    detailedDescriptions: DetailedDescription[];
    eerCurrentRating: string;
    eerPotentialRating: string;
    eirCurrentRating: string;
    eirPotentialRating: string;
    featureProperty: boolean;
    featureList: string[];
    floorLevels: string[];
    floors: number;
    furnishedState: FurnishedState;
    groundRent: number;
    groundRentReviewPeriod: string;
    leaseExpiry: string;
    lastModified: Date;
    lifeCycleStatus: LifeCycleStatus;
    listedBuilding: boolean;
    listedBuildingGrade: ListedBuildingGrade;
    listingMedia: ListingMediaDto[];
    listingReference: string;
    listingLocation: ListingLocation;
    livingRooms: number;
    minimumContractLength: number;
    minimumContractLengthUnits: UnitOfTime;
    newBuild: boolean;
    openDay: Date;
    parking: boolean;
    petsAllowed: boolean;
    pricing: Pricing;
    propertyType: PropertyType;
    rateableValue: number;
    ratesPayable: number;
    rentalTerm: RentalTerm;
    repossession: boolean;
    retirement: boolean;
    sapRating: string;
    serviceCharge: ServiceCharge;
    serviced: boolean;
    sourceUri: string;
    sharedAccommodation: boolean;
    sharedOwnershipDetails: string;
    smokersConsidered: boolean;
    spiderTag: string;
    summaryDescription: string;
    tenanted: boolean;
    tenantEligibilityDss: Eligibility;
    tenantEligibilityStudents: Eligibility;
    tenure: Tenure;
    totalBedrooms: number;
    whiteGoods: string[];
    watchers: WatcherDto[];
    keyContacts: KeyContact[];
}

export class Listing implements Listing {
    constructor(initialValues?: ListingFormValues) {
        Object.assign(this, initialValues);
    }
}

export class ListingFormValues {
    id?: string = '';
    accessibility?: boolean = undefined;
    accessStatus: AccessStatus = AccessStatus.Public;
    acreage?: number;
    agency?: string;
    addedOn: Date = new Date();
    administrationFees?: string = '';
    annualBusinessRates?: number = undefined;
    areaTotal?: number = undefined;
    areaUnits: UnitOfArea = UnitOfArea.SqMetres;
    availableBedrooms?: number = undefined;
    availableFromDate?: Date = undefined;
    bathrooms?: number = undefined;
    billsIncluded?: string[] = [];
    businessForSale?: boolean = undefined;
    buyerIncentives?: string[] = [];
    category?: Category = Category.residential;
    centralHeating?: CentralHeating = undefined;
    chainFree: boolean = true;
    companyId: string = '';
    commercialUseClass?: string[] = [];
    commonholdDetails: string = ''
    connectedUtilities?: string[] = [];
    constructionYear?: number = undefined;
    cookerType?: CookerType = undefined;
    councilTaxBand?: CouncilTaxBand = undefined;
    decorativeCondition?: DecorativeCondition = undefined;
    deposit?: number = undefined;
    detailedDescriptions?: DetailedDescription[] = [];
    eerCurrentRating?: string = '';
    eerPotentialRating?: string = '';
    eirCurrentRating?: string = '';
    eirPotentialRating?: string = '';
    featureProperty?: boolean = false;
    featureList?: string[] = [];
    floorLevels?: string[] = [];
    floors?: number = undefined;
    furnishedState?: FurnishedState = undefined;
    groundRent?: number = undefined;
    groundRentReviewPeriod?: string = '';
    leaseExpiry?: string = '';
    lifeCycleStatus?: LifeCycleStatus = undefined;
    listedBuilding?: boolean = false;
    listedBuildingGrade?: ListedBuildingGrade = undefined;
    listingMedia: ListingMediaDto[] = [];
    listingReference?: string = '';
    listingLocation: ListingLocation = {
        id: '',
        displayAddress: '',
        propertyNumberOrName: '',
        streetName: '',
        locality: '',
        townOrCity: '',
        county: '',
        postalCode: '',
        country: 'United Kingdom',
        latitude: 51.505,
        longitude: -0.09,
    };
    livingRooms?: number = undefined;
    minimumContractLength?: number = undefined;
    minimumContractLengthUnits?: UnitOfTime = UnitOfTime.Months;
    newBuild?: boolean = undefined;
    openDay?: Date = undefined;
    parking?: boolean = true;
    petsAllowed?: boolean = undefined;
    pricing: Pricing = {
        id: '',
        transactionType: TransactionType.rent,
        currency: "gbp",
        price: undefined,
        pricePerUnitArea: undefined,
        rentFrequency: Frequency.perMonth,
        priceQualifier: "",
        auction: false,
        areaUnits: UnitOfArea.SqMetres
    };
    propertyType?: PropertyType = undefined;
    rateableValue?: number = undefined;
    ratesPayable?: number = undefined;
    rentalTerm?: RentalTerm = undefined;
    repossession?: boolean = undefined;
    retirement?: boolean = undefined;
    sapRating?: string = '';
    serviceCharge?: ServiceCharge = undefined;
    serviced?: boolean = undefined;
    sourceUri?: string;
    sharedAccommodation?: boolean = undefined;
    sharedOwnershipDetails?: string = '';
    smokersConsidered?: boolean = undefined;
    summaryDescription?: string = '';
    tenanted?: boolean = undefined;
    tenantEligibilityDss?: Eligibility = undefined;
    tenantEligibilityStudents?: Eligibility = undefined;
    tenure?: Tenure = 1;
    totalBedrooms?: number = undefined;
    whiteGoods?: string[] = [];

    constructor(listing?: ListingFormValues) {
        if (listing) {
            this.id = listing.id;
            this.accessibility = listing.accessibility;
            this.accessStatus = listing.accessStatus;
            this.acreage = listing.acreage;
            this.agency = listing.agency;
            this.addedOn = listing.addedOn;
            this.administrationFees = listing.administrationFees;
            this.annualBusinessRates = listing.annualBusinessRates;
            this.areaTotal = listing.areaTotal;
            this.areaUnits = listing.areaUnits;
            this.availableBedrooms = listing.availableBedrooms;
            this.availableFromDate = listing.availableFromDate;
            this.bathrooms = listing.bathrooms;
            this.billsIncluded = listing.billsIncluded;
            this.businessForSale = listing.businessForSale;
            this.buyerIncentives = listing.buyerIncentives;
            this.category = listing.category;
            this.centralHeating = listing.centralHeating;
            this.chainFree = listing.chainFree;
            this.companyId = listing.companyId;
            this.commercialUseClass = listing.commercialUseClass;
            this.commonholdDetails = listing.commonholdDetails;
            this.connectedUtilities = listing.connectedUtilities;
            this.constructionYear = listing.constructionYear;
            this.cookerType = listing.cookerType;
            this.councilTaxBand = listing.councilTaxBand;
            this.decorativeCondition = listing.decorativeCondition;
            this.deposit = listing.deposit;
            this.detailedDescriptions = listing.detailedDescriptions;
            this.eerCurrentRating = listing.eerCurrentRating;
            this.eerPotentialRating = listing.eerPotentialRating;
            this.eirCurrentRating = listing.eirCurrentRating;
            this.eirPotentialRating = listing.eirPotentialRating;
            this.featureProperty = listing.featureProperty;
            this.featureList = listing.featureList;
            this.floorLevels = listing.floorLevels;
            this.floors = listing.floors;
            this.furnishedState = listing.furnishedState;
            this.groundRent = listing.groundRent;
            this.groundRentReviewPeriod = listing.groundRentReviewPeriod;
            this.leaseExpiry = listing.leaseExpiry;
            this.lifeCycleStatus = listing.lifeCycleStatus;
            this.listedBuilding = listing.listedBuilding;
            this.listedBuildingGrade = listing.listedBuildingGrade;
            this.listingMedia = listing.listingMedia;
            this.listingReference = listing.listingReference;
            this.listingLocation = listing.listingLocation;
            this.livingRooms = listing.livingRooms;
            this.minimumContractLength = listing.minimumContractLength;
            this.minimumContractLengthUnits = listing.minimumContractLengthUnits;
            this.newBuild = listing.newBuild;
            this.openDay = listing.openDay;
            this.parking = listing.parking;
            this.petsAllowed = listing.petsAllowed;
            this.pricing = listing.pricing;
            this.propertyType = listing.propertyType;
            this.rateableValue = listing.rateableValue;
            this.ratesPayable = listing.ratesPayable;
            this.rentalTerm = listing.rentalTerm;
            this.repossession = listing.repossession;
            this.retirement = listing.retirement;
            this.sapRating = listing.sapRating;
            this.serviceCharge = listing.serviceCharge;
            this.serviced = listing.serviced;
            this.sourceUri = listing.sourceUri;
            this.sharedAccommodation = listing.sharedAccommodation;
            this.sharedOwnershipDetails = listing.sharedOwnershipDetails;
            this.smokersConsidered = listing.smokersConsidered;
            this.summaryDescription = listing.summaryDescription;
            this.tenanted = listing.tenanted;
            this.tenantEligibilityDss = listing.tenantEligibilityDss;
            this.tenantEligibilityStudents = listing.tenantEligibilityStudents;
            this.tenure = listing.tenure;
            this.totalBedrooms = listing.totalBedrooms;
            this.whiteGoods = listing.whiteGoods;
        }
    }
}



